### PR TITLE
feat(pagination): implement component

### PR DESCRIPTION
**Pull request by Leotheluck**
> Pagination is the process of splitting information over multiple pages instead of showing it all on a single page; also the name for the interface component used for navigating between these pages.

### CODE REVIEW

**Review by astagnol — CHANGES_REQUESTED** [review]
- We need to add style (P-800 color) to button in current state. Check former Pagination component
- wrong focus behaviour. For example: go to Demo -> 'tab' to page 6 -> 'enter' -> 'tab' again = focus restarts from the previous button

File: packages/ods/src/components/pagination/src/components/ods-pagination/ods-pagination.tsx [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+    if (this.current < this.pageList.length) {
+      this.onKeyDown(event, page + 1);
+    }
+  }
+
+  handlePageKeyDown(event: KeyboardEvent, page: number): void {
+    this.onKeyDown(event, page);
+  }
+
+  onKeyDown(event: KeyboardEvent, page: number): void {
```
Comment (dpellier) [inline packages/ods/src/components/pagination/src/components/ods-pagination/ods-pagination.tsx:184]: if possible use `onKeyUp` (and manage the Space prevention to avoid scroll on key down)

File: packages/ods/src/components/pagination/src/components/ods-pagination/ods-pagination.tsx [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+    if (event.code === 'Enter' || event.code === 'Space') {
```
Comment (dpellier) [inline packages/ods/src/components/pagination/src/components/ods-pagination/ods-pagination.tsx:185]:
```
(event.key === ' ' || event.key === 'Enter')
```

File: packages/ods/src/components/pagination/src/components/ods-pagination/ods-pagination.tsx [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+      event.preventDefault();
+      this.setPageIndex(page);
+    }
+  }
+
+  renderArrows(direction: 'left' | 'right'): typeof Fragment {
```
Comment (dpellier) [inline packages/ods/src/components/pagination/src/components/ods-pagination/ods-pagination.tsx:191]: `renderArrow`

File: packages/ods/src/components/pagination/src/components/ods-pagination/ods-pagination.tsx [hunk 4]
```diff
@@ -0,0 +1,338 @@
...
+      <li class="ods-pagination__list__arrow">
+        <ods-button
+          color={ODS_BUTTON_COLOR.primary}
+          class="ods-pagination__list__arrow__button"
+          icon={isLeft ? ODS_ICON_NAME.chevronLeft : ODS_ICON_NAME.chevronRight}
+          id={arrowButtonId}
+          isDisabled={this.isDisabled || (isLeft && this.current === 1) || (!isLeft && this.current >= this.pageList.length)}
+          label=""
+          onClick={() => {
+            if (this.isDisabled || (isLeft && this.current === 1) || (!isLeft && this.current >= this.pageList.length)) {
```
Comment (dpellier) [inline packages/ods/src/components/pagination/src/components/ods-pagination/ods-pagination.tsx:206]: add this logic to a `const` as it is duplicated

File: packages/ods/src/components/pagination/src/components/ods-pagination/ods-pagination.tsx [hunk 5]
```diff
@@ -0,0 +1,338 @@
...
+
+          {
+            this.pageList
+              .filter((page) => page.active)
+              .map((page) => {
+                const pageId = this.pageList.indexOf(page) + 1;
+
+                this.pageList.length > 6 && this.pageList.length - this.current > 3 && pageId === this.pageList.length && this.renderEllipsis();
+
+                this.pageList.length > 6 && this.current > 4 && pageId === 1 && this.renderEllipsis();
```
Comment (dpellier) [inline packages/ods/src/components/pagination/src/components/ods-pagination/ods-pagination.tsx:303]: replace those two inline with a single function that manages both cases
\+ use global constants instead of magic numbers

File: packages/ods/src/components/pagination/src/components/ods-pagination/ods-pagination.tsx [hunk 6]
```diff
@@ -0,0 +1,338 @@
...
+
+                return (
+                  <div class="ods-pagination__list__page">
+                    { this.pageList.length > 6 && this.pageList.length - this.current > 3 && pageId === this.pageList.length && this.renderEllipsis() }
```
Comment (dpellier) [inline packages/ods/src/components/pagination/src/components/ods-pagination/ods-pagination.tsx:307]: and use the result of the function here

File: packages/ods/src/components/pagination/src/components/ods-pagination/ods-pagination.tsx [hunk 7]
```diff
@@ -0,0 +1,338 @@
...
+                        class={{
+                          'ods-pagination__list__page__button': true,
+                          'ods-pagination__list__page__button--selected': this.current === pageId,
+                        }}
+                        variant={this.current === pageId ? ODS_BUTTON_VARIANT.default : ODS_BUTTON_VARIANT.ghost}
+                        isDisabled={this.isDisabled}
+                        label={`${pageId}`}
+                        color={ODS_BUTTON_COLOR.primary}
+                        size={ODS_BUTTON_SIZE.md}
+                        onClick={(): void => this.handlePageClick(Number(pageId))}
```
Comment (dpellier) [inline packages/ods/src/components/pagination/src/components/ods-pagination/ods-pagination.tsx:321]: why the extra `Number` cast?

File: packages/ods/src/components/pagination/src/constants/pagination-per-page.ts [hunk 8]
```diff
@@ -0,0 +1,20 @@
+enum ODS_PAGINATION_PER_PAGE {
+  min = 10,
```
Comment (dpellier) [inline packages/ods/src/components/pagination/src/constants/pagination-per-page.ts:2]: TBD not sur the `min`, `max` value is that great here, would be much more consistent to have same defined value for each, like `option_10`

**Review by dpellier — CHANGES_REQUESTED** [review]

File: packages/ods/src/components/pagination/src/components/ods-pagination/ods-pagination.tsx [hunk 9]
```diff
@@ -0,0 +1,329 @@
...
+
+    if (this.current === 1) {
+      // If current is already 1 we don't want to emit a change event
+      this.updatePageList();
+    } else {
+      await this.setCurrentPage(1);
+    }
+  }
+
+  handlePreviousClick(page: number): void {
```
Comment (dpellier) [inline packages/ods/src/components/pagination/src/components/ods-pagination/ods-pagination.tsx:143]: set all methods `private`

File: packages/ods/src/components/pagination/src/components/ods-pagination/ods-pagination.tsx [hunk 10]
```diff
@@ -0,0 +1,329 @@
...
+      this.setCurrentPage(page);
+    }
+  }
+
+  renderArrow(direction: 'left' | 'right'): typeof Fragment {
+    const isLeft = direction === 'left';
+    const tooltipLabel = isLeft ? this.labelTooltipPrevious : this.labelTooltipNext;
+    const arrowButtonId = isLeft ? this.leftArrowButtonId : this.rightArrowButtonId;
+
+    const checkArrowDisabled = (): boolean => {
```
Comment (dpellier) [inline packages/ods/src/components/pagination/src/components/ods-pagination/ods-pagination.tsx:183]: why a function?
it will be called multiple times on each render for the same result each time

File: packages/ods/src/components/pagination/src/components/ods-pagination/ods-pagination.scss [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+        justify-content: center;
+        width: 2.5rem;
+      }
+    }
+
+    &__page {
+      &__ellipsis {
+        display: flex;
+        justify-content: center;
+        cursor: not-allowed;
```
Comment (dpellier) [inline packages/ods/src/components/pagination/src/components/ods-pagination/ods-pagination.scss:57]: should be managed by the ods-button style already, otherwise we may not want to use an ods-button for a disabled element

File: packages/ods/src/components/pagination/src/components/ods-pagination/ods-pagination.scss [hunk 12]
```diff
@@ -0,0 +1,69 @@
+:host(.ods-pagination) {
+  display: flex;
+  flex-direction: row;
+  align-items: center;
+
+  li {
```
Comment (dpellier) [inline packages/ods/src/components/pagination/src/components/ods-pagination/ods-pagination.scss:6]: try to avoid generic selector + the style is duplicated on `__results` and `__list`

File: packages/ods/src/components/pagination/src/components/ods-pagination/ods-pagination.tsx [hunk 13]
```diff
@@ -0,0 +1,329 @@
...
+        <ul class="ods-pagination__list">
+          { this.renderArrow( 'left' ) }
+
+          {
+            this.pageList
+              .filter((page) => page.active)
+              .map((page) => {
+                const pageId = this.pageList.indexOf(page) + 1;
+                const shouldRenderLeftEllipsis = this.pageList.length > 6 && this.pageList.length - this.current > 3 && pageId === this.pageList.length;
+                const shouldRenderRightEllipsis = this.pageList.length > 6 && this.current > 4 && pageId === 1;
```
Comment (dpellier) [inline packages/ods/src/components/pagination/src/components/ods-pagination/ods-pagination.tsx:292]: use global constants instead of magic numbers

**Review by dpellier — CHANGES_REQUESTED** [review]

**Review by dpellier — CHANGES_REQUESTED** [review]
Capture d’écran 2024-07-16 à 09 46 28

z-index issue on focus

**Review by dpellier — CHANGES_REQUESTED** [review]
Issue on demo `Total Items`
-> change select value to 100
=> 
Capture d’écran 2024-07-16 à 09 48 49

**Review by feoche — CHANGES_REQUESTED** [review]
## Global
- [x] When focusing "Previous"/"Next" Buttons, page number is decremented/incremented twice
- [x] Setting `totalPages` to 8 when current page is 4 breaks the component
- [x] Tweaking `totalItems` and items per page breaks the behavior/amount of pages
- [x] Setting a big number (>100 000) of totalItems freezes the browser

## Style
- [x] Keep current page Button text color `white` when Pagination is disabled

## Documentation
- [x] Remove "Default" example
- [x] Create "Default" with `totalPages=7` and `defaultCurrentPage=4`
- [x] Rename "Default Current Page" to "Default - Large amount of pages"
- [x] Rename "Tooltips" to "Tooltip - Configuration"
- [x] Add "Total items with labels" example
- [x] Remove "Total pages", as it is included in previous examples

File: packages/ods/src/components/pagination/src/components/ods-pagination/ods-pagination.tsx [hunk 14]
```diff
@@ -0,0 +1,396 @@
...
+                <slot name="after-total-items"></slot>
+              </ods-text>
+            </div>
+        }
+
+        <ul class="ods-pagination__list">
+          {this.renderArrow('left')}
+
+          {(this.totalItems || this.actualTotalPages > 1) && (
+            <li key={1}>
```
Comment (aesteves60) [inline packages/ods/src/components/pagination/src/components/ods-pagination/ods-pagination.tsx:325]: this part can be mutualize in a function there are just the `key` and the `ods-pagination__list__page__button--visible` class who change

File: packages/ods/src/components/pagination/src/controller/ods-pagination.ts [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+  }
+
+  return Math.ceil(totalItems / (itemPerPage || 1));
+}
+
+function createPageList(totalPages: number, pageSelected: number): OdsPaginationPageList {
+  const pageList: OdsPaginationPageList = [];
+
+  // Create initial pageList with 'active' property set to false for each page.
+  for (let i = 1; i <= totalPages; i++) {
```
Comment (aesteves60) [inline packages/ods/src/components/pagination/src/controller/ods-pagination.ts:15]: `const pageList = new Array(totalPages).fill({ active: false, isVisible: false })`

File: packages/ods/src/components/pagination/src/components/ods-pagination/ods-pagination.tsx [hunk 16]
```diff
@@ -0,0 +1,396 @@
...
+      } else if (this.current <= this.ellipsisThreshold) {
+        page.isVisible = pageId <= maxVisibleItems - 2;
+      } else if (this.current >= this.actualTotalPages - this.ellipsisThreshold) {
+        page.isVisible = pageId >= this.actualTotalPages - (maxVisibleItems - 3);
+      } else {
+        page.isVisible = pageId >= this.current - 1 && pageId <= this.current + 1;
+      }
+    });
+
+    const visiblePages = this.pageList.filter((page) => page.isVisible);
```
Comment (aesteves60) [inline packages/ods/src/components/pagination/src/components/ods-pagination/ods-pagination.tsx:150]: I think this part is not used because you don't edit any component's variable

File: packages/ods/src/components/pagination/src/components/ods-pagination/ods-pagination.tsx [hunk 17]
```diff
@@ -0,0 +1,396 @@
...
+
+  private updatePageList(): void {
+    this.pageList = createPageList(this.actualTotalPages, this.current).map((page) => ({
+      ...page,
+      isVisible: false,
+    }));
+  }
+
+  private updatePageVisibility(): void {
+    const maxVisibleItems = 7;
```
Comment (aesteves60) [inline packages/ods/src/components/pagination/src/components/ods-pagination/ods-pagination.tsx:132]: Declare it in the top of the component

**Review by aesteves60 — CHANGES_REQUESTED** [review]

File: packages/ods/src/components/pagination/src/components/ods-pagination/ods-pagination.scss [hunk 18]
```diff
@@ -0,0 +1,82 @@
...
+        width: 2.5rem;
+      }
+    }
+
+    &__page {
+      &__ellipsis {
+        display: flex;
+        justify-content: center;
+        width: 2.5rem;
+        pointer-events: auto;
```
Comment (dpellier) [inline packages/ods/src/components/pagination/src/components/ods-pagination/ods-pagination.scss:53]: ?

File: packages/ods/src/components/pagination/src/components/ods-pagination/ods-pagination.scss [hunk 19]
```diff
@@ -0,0 +1,82 @@
...
+      align-items: center;
+    }
+
+    &__arrow {
+      padding: 0 8px;
+
+      &__button::part(button) {
+        display: flex;
+        justify-content: center;
+        width: 2.5rem;
```
Comment (dpellier) [inline packages/ods/src/components/pagination/src/components/ods-pagination/ods-pagination.scss:44]: could be nice to use a sass var to share this value everywhere here `$ods-pagination-button-width`

File: packages/ods/src/components/pagination/src/components/ods-pagination/ods-pagination.tsx [hunk 20]
```diff
@@ -0,0 +1,331 @@
...
+  async onTotalPagesChange(): Promise<void> {
+    if (!this.totalItems) {
+      this.actualTotalPages = this.totalPages;
+    }
+    this.updatePageList();
+  }
+
+  @Watch('itemPerPage')
+  async onItemPerPageChange(): Promise<void> {
+    if (!this.isFirstLoad) {
```
Comment (dpellier) [inline packages/ods/src/components/pagination/src/components/ods-pagination/ods-pagination.tsx:79]: Watch are not supposed to be executed on initial load, did you try to remove it to check if it was still needed?

**Review by dpellier — CHANGES_REQUESTED** [review]
Controller tests are missing.
It's quite the most important as the `createPageList` as lots of edge cases to validate


Tooltips are still visible when disabled
Capture d’écran 2024-07-17 à 16 26 18

File: packages/ods/src/components/pagination/src/components/ods-pagination/ods-pagination.tsx [hunk 21]
```diff
@@ -0,0 +1,327 @@
...
+  shadow: true,
+  styleUrl: 'ods-pagination.scss',
+  tag: 'ods-pagination',
+})
+export class OdsPagination {
+  private actualTotalPages = 1;
+  private leftArrowButtonId = 'pagination-left-arrow';
+  private rightArrowButtonId = 'pagination-right-arrow';
+  private hostId: string = '';
+  private MAX_VISIBLE_ITEMS = 7;
```
Comment (dpellier) [inline packages/ods/src/components/pagination/src/components/ods-pagination/ods-pagination.tsx:23]: Do not redeclare the constants, use the one from the controller

File: packages/ods/src/components/pagination/src/components/ods-pagination/ods-pagination.scss (hunk 19 above)
Comment (dpellier) [reply packages/ods/src/components/pagination/src/components/ods-pagination/ods-pagination.scss:46]: use `pagination.$ods-pagination-button-width;`

File: packages/ods/src/components/pagination/src/controller/ods-pagination.ts [hunk 22]
```diff
@@ -0,0 +1,80 @@
...
+  }
+
+  return Math.ceil(totalItems / (itemPerPage || 1));
+}
+
+function createPageList(totalPages: number, pageSelected: number): OdsPaginationPageList {
+  const MAX_VISIBLE_ITEMS = 7;
+  const ELLIPSIS_THRESHOLD = 4;
+  const MINIMUM_PAGE = 1;
+  const DEFAULT_PAGE_OFFSET = 2;
```
Comment (dpellier) [inline packages/ods/src/components/pagination/src/controller/ods-pagination.ts:15]: declare the constants outside of any functions, export them at the end of the file and use them in the component

File: packages/ods/src/components/pagination/tests/controller/ods-pagination.spec.ts [hunk 23]
```diff
@@ -0,0 +1,195 @@
...
+
+    it('should activate appropriate pages when totalPages is greater than MAX_VISIBLE_ITEMS and pageSelected is in the middle', () => {
+      const totalPages = 10;
+      const pageSelected = 5;
+
+      const result = createPageList(totalPages, pageSelected);
+
+      expect(result).toHaveLength(totalPages);
+      result.forEach((page, index) => {
+        if (index === 0 || index === totalPages - 1 || (index >= (pageSelected - 2) && index <= pageSelected)) {
```
Comment (dpellier) [inline packages/ods/src/components/pagination/tests/controller/ods-pagination.spec.ts:87]: Do not rewrite logic in unit tests (otherwise you should test it also)
use a static call for each expected results
```
expect(result[0].active).toBe(...
expect(result[1].active).toBe(...
```

Comment (dpellier) [reply packages/ods/src/components/pagination/tests/controller/ods-pagination.spec.ts:87]: (same for all tests where you have some logic)

File: packages/ods/src/style/_pagination.scss [hunk 24]
```diff
@@ -0,0 +1 @@
+$ods-pagination-button-width: 2.5rem;
```
Comment (dpellier) [inline packages/ods/src/style/_pagination.scss:1]: you can declare it directly in the component

**Review by dpellier — CHANGES_REQUESTED** [review]

**Review by feoche — APPROVED** [review]

**Review by dpellier — APPROVED** [review]

**Review by astagnol — APPROVED** [review]

**Review by aesteves60 — APPROVED** [review]
